src/test/fuzz/p2p_transport_serialization.cpp

```diff
@@ -63,22 +65,260 @@ FUZZ_TARGET(p2p_transport_serialization, .init = initialize_p2p_transport_serial
+        return send_now > 0;
+    };
+
+    // Function to make !side receive bytes (if any).
+    auto recv_fn = [&](int side, bool everything = false) {
+        // Don't do anything if no bytes in flight.
+        if (in_flight[side].empty()) return false;
+        // Decide span to receive
+        size_t to_recv_len = in_flight[side].size();
+        if (!everything) to_recv_len = provider.ConsumeIntegralInRange<size_t>(0, to_recv_len);
+        Span<const uint8_t> to_recv = Span{in_flight[side]}.first(to_recv_len);
+        // Process those bytes
+        while (!to_recv.empty()) {
+            size_t old_len = to_recv.size();
+            bool ret = transports[!side]->ReceivedBytes(to_recv);
+            // Bytes must always be accepted, as this test does not introduce any errors in
+            // communication.
+            assert(ret);
+            // Clear cached expected 'more' information: if certainly no more data was to be sent
+            // before, receiving bytes makes this uncertain.
+            if (expect_more[!side] == false) expect_more[!side] = std::nullopt;
+            // Verify consistency of GetBytesToSend after ReceivedBytes
+            bytes_to_send_fn(!side);
+            bool progress = to_recv.size() < old_len;
+            if (transports[!side]->ReceivedMessageComplete()) {
+                bool reject{false};
+                auto received = transports[!side]->GetReceivedMessage({}, reject);
+                // Receiving must succeed.
+                assert(!reject);
+                // There must be a corresponding expected message.
+                assert(!expected[side].empty());
+                // The m_message_size field must be correct.
+                assert(received.m_message_size == received.m_recv.size());
+                // The m_type must match what is expected.
+                assert(received.m_type == expected[side].front().m_type);
+                // The data must match what is expected.
+                assert(MakeByteSpan(received.m_recv) == MakeByteSpan(expected[side].front().data));
+                expected[side].pop_front();
+                progress = true;
+            }
+            // Progress must be made (by processing incoming bytes and/or returning complete
+            // messages) until all received bytes are processed.
+            assert(progress);
+        }
+        // Remove the processed bytes from the in_flight buffer.
+        in_flight[side].erase(in_flight[side].begin(), in_flight[side].begin() + to_recv_len);
+        // Return whether anything was received.
+        return to_recv_len > 0;
+    };
+
+    // Main loop, interleaving new messages, sends, and receives.
+    unsigned iter = 0;
+    while (iter < 1000 && provider.remaining_bytes()) {
+        uint8_t code = provider.ConsumeIntegral<uint8_t>();
+        // Lowest bit identifies the sender side of the operation.
+        int side = code & 1;
+        switch ((side >> 1) % 3) {
```

Comment:
Given that `side` is `0` or `1`, then `side >> 1` will always be `0`, no? Did you mean `code >> 1`? I find this less cryptic:

```cpp
size_t side = provider.ConsumeBool();
switch (provider.ConsumeIntegralInRange(0, 2)) {
    case 0: ...
    case 1: ...
    case 2: ...
}